backend/src/main/java/org/eclipse/tractusx/puris/backend/demand/logic/adapter/ShortTermMaterialDemandSammMapper.java

```diff
@@ -0,0 +1,160 @@
+        if (demandList == null || demandList.isEmpty()) {
+            log.warn("Can't map empty list");
+            return null;
+        }
+        Partner partner = demandList.get(0).getPartner();
+        Material material = demandList.get(0).getMaterial();
+        if (demandList.stream().anyMatch(dem -> !dem.getPartner().equals(partner))) {
+            log.warn("Can't map demand list with different partners");
+            return null;
+        }
+
+        if (demandList.stream().anyMatch(prod -> !prod.getMaterial().equals(material))) {
+            log.warn("Can't map demand list with different materials");
+            return null;
+        }
+        var groupedByCategory = demandList
+                .stream()
+                .collect(Collectors.groupingBy(demand -> new DemandGroupingHelper(demand.getDemandCategoryCode(), demand.getDemandLocationBpns(), demand.getSupplierLocationBpns())));
+        ShortTermMaterialDemand samm = new ShortTermMaterialDemand();
+
+        var mpr = mprService.findAll().stream().filter(mr -> mr.getMaterial().equals(material) && mr.getPartner().equals(partner)).findFirst().orElse(null);
+        if (mpr == null) {
+            log.warn("Could not identify materialPartnerRelation with ownMaterialNumber " + material.getOwnMaterialNumber()
+                    + " and partner bpnl " + partner.getBpnl());
+            return null;
+        }
+        samm.setMaterialGlobalAssetId(mpr.getPartnerCXNumber());
+
+        var demandSeriesList = new HashSet<DemandSeries>();
+        samm.setDemandSeries(demandSeriesList);
+        for (var mappingHelperListEntry : groupedByCategory.entrySet()) {
+            var key = mappingHelperListEntry.getKey();
+            DemandSeries demandSeries = new DemandSeries();
+            demandSeriesList.add(demandSeries);
+            demandSeries.setLastUpdatedOnDateTime(new Date());
+            demandSeries.setDemandCategory(mapDemandCategory(key.category()));
+            demandSeries.setCustomerLocationBpns(key.customerLocationBpns());
+            demandSeries.setExpectedSupplierLocationBpns(key.expectedSupplierLocationBpns());
+            var demands = new HashSet<Demand>();
+            demandSeries.setDemands(demands);
+            for (var v : mappingHelperListEntry.getValue()) {
+                ItemQuantityEntity itemQuantityEntity = new ItemQuantityEntity(v.getQuantity(), v.getMeasurementUnit());
+                Demand dailyDemand = new Demand(itemQuantityEntity, v.getDay());
+                demands.add(dailyDemand);
+            }
+        }
+        return samm;
+    }
+
+    public List<ReportedDemand> sammToReportedDemand(ShortTermMaterialDemand samm, Partner partner) {
+        log.info("Mapping SAMM: " + samm);
+        String matNbrCatenaX = samm.getMaterialGlobalAssetId();
+        ArrayList<ReportedDemand> outputList = new ArrayList<>();
+
+        var material = materialService.findByMaterialNumberCx(matNbrCatenaX);
```

Comment:
Please use `mprService.findAllByPartnerAndAndPartnerCXNumber(partner, matNbrCatenax)`

Reply:
I thought since Short Term Material Demand is requested by the supplier, the supplier should use their own cx number rather than the partner's. From the customer's perspective the Demand is sent using the partner's cx number.

Reply:
Catena-X number handling is maybe a little bit strange:
- we always generate the numbers on material creation
- If we're acting as supplier, this number is put in the ShellDescriptor of the twin
- If we're acting as a customer, we more or less 
  - search for the ShellDescriptor of the twin at the partner side
  - pull the partTypeInformation Submodel and extract the CX ID
  - remember that information in the mpr

Reason for that behavior is that a supplier creates the twin and the customer uses this cx id as the leading one. In Multisourcing a customer creates a twin per supplier - a syncronization of the cx ids is neither wanted from business perspective nor possible due to antitrust

Reply:
Nevermind, I was not thinking from the correct point of view.